man/coap_encryption.txt.in

```diff
@@ -580,13 +878,212 @@ setup_server_context_pki (const char *public_cert_file,
+
+typedef struct valid_psk_snis_t {
+  int count;
+  sni_psk_def_t *sni_list;
+} valid_psk_snis_t;
+
+/*
+ * PSK Subject Name Identifier (SNI) callback verifier
+ */
+static const coap_dtls_spsk_info_t *
+verify_psk_sni_callback(coap_str_const_t *sni,
+                        coap_session_t *c_session,
+                        void *arg
+) {
+  valid_psk_snis_t *valid_sni_list = (valid_psk_snis_t *)arg;
+  int i;
+  /* Remove (void) definition if variable is used */
+  (void)c_session;
+
+  /* Check that the SNI is valid */
+  for (i = 0; i < valid_sni_list->count; i++) {
+    if (!strcasecmp((const char*)sni->s, valid_sni_list->sni_list[i].sni)) {
+      return &valid_sni_list->sni_list[i].psk_info;
+    }
+  }
+  return NULL;
+}
+
+static coap_context_t *
+setup_server_context_psk (const char *hint,
+                          const uint8_t *key,
+                          unsigned int key_len,
+                          valid_ids_t *valid_id_list,
+                          valid_psk_snis_t *valid_sni_list
+) {
+  coap_endpoint_t *endpoint;
+  coap_address_t listen_addr;
+  coap_context_t *context;
+  coap_dtls_spsk_t dtls_psk;
+
+  /* See coap_tls_library(3) */
+  if (!coap_dtls_is_supported())
+    return NULL;
+
+  context = coap_new_context(NULL);
+  if (!context)
+    return NULL;
+
+  memset (&dtls_psk, 0, sizeof (dtls_psk));
+
+  /* see coap_encryption(3) */
+  dtls_psk.version                 = COAP_DTLS_SPSK_SETUP_VERSION;
+  dtls_psk.validate_id_call_back   = verify_id_callback;
+  dtls_psk.id_call_back_arg        = valid_id_list;
+  dtls_psk.validate_sni_call_back  = verify_psk_sni_callback;
+  dtls_psk.sni_call_back_arg       = valid_sni_list;
+  dtls_psk.psk_info.hint.s         = (const uint8_t*)hint;
+  dtls_psk.psk_info.hint.length    = hint ? strlen(hint) : 0;
+  dtls_psk.psk_info.key.s          = key;
+  dtls_psk.psk_info.key.length     = key_len;
+
+  if (coap_context_set_psk2(context, &dtls_psk)) {
+    coap_free_context(context);
+    return NULL;
+  }
+
+  coap_address_init(&listen_addr);
+  listen_addr.addr.sa.sa_family = AF_INET;
+  listen_addr.addr.sin.sin_port = htons (5684);
+
+  endpoint = coap_new_endpoint(context, &listen_addr, COAP_PROTO_DTLS);
+  if (!endpoint) {
+    coap_free_context(context);
+    return NULL;
+  }
+
+  /* Initialize resources - See coap_resource(3) init_resources() example */
+
+  return context;
+}
+----
+
+*CoAP Client DTLS PSK Setup*
+[source, c]
+----
+#include <coap@LIBCOAP_API_VERSION@/coap.h>
+
+#include <stdio.h>
+
+#ifndef min
+#define min(a,b) ((a) < (b) ? (a) : (b))
+#endif
+
+static const coap_dtls_cpsk_info_t *
+verify_ih_callback(coap_str_const_t *hint,
+                   coap_session_t *c_session,
+                   void *arg
+) {
+  coap_dtls_cpsk_info_t *psk_info = (coap_dtls_cpsk_info_t *)arg;
+  char lhint[128];
+  /* Remove (void) definition if variable is used */
+  (void)c_session;
+
+  snprintf(lhint, sizeof(lhint), "%.*s", (int)hint->length, hint->s);
```

Comment:
Pass `"%.*s"` to `coap_log()`, see above.

Reply:
Updated